src/js/player.js

```diff
@@ -823,6 +825,179 @@ class Player extends Component {
+   * the source is changed. Fires `sourceset` just after the source has changed
+   *
+   * @fires Player#sourceset
+   */
+  watchForSourceSet_() {
+    // if we cannot overwrite the src property, there is no support
+    // iOS 7 safari for instance cannot do this.
+    try {
+      const el = document.createElement('video');
+
+      Object.defineProperty(el, 'src', {
+        get() {},
+        set() {}
+      });
+    } catch (e) {
+      return;
+    }
+
+    const el = this.el().getElementsByTagName('video')[0] || this.el().getElementsByTagName('audio')[0];
+
+    if (!el) {
+      if (!this.isReady_) {
+        this.ready(() => this.watchForSourceSet_());
+      }
+      return;
+    }
+
+    // we need to fire sourceset when the player is ready
+    // if we find that the media element had a src when it was
+    // given to us
+    if (this.tagAttributes && this.tagAttributes.src) {
+      this.ready(() => {
+        this.trigger('sourceset');
+      });
+    }
+
+    const proto = window.HTMLMediaElement.prototype;
+    const srcDescriptor = {
+      get() {
+        return proto.getAttribute.call(this, 'src');
+      },
+      set(v) {
+        return proto.setAttribute.call(this, 'src', v);
+      },
+      enumerable: true
+    };
+
+    // preserve getters/setters already on `el.src` if they exist
+    if (Object.getOwnPropertyDescriptor(el, 'src')) {
+      Object.assign(srcDescriptor, Object.getOwnPropertyDescriptor(el, 'src'));
+    } else if (Object.getOwnPropertyDescriptor(proto, 'src')) {
+      Object.assign(srcDescriptor, Object.getOwnPropertyDescriptor(proto, 'src'));
+    }
+
+    Object.defineProperty(el, 'src', {
+      get: srcDescriptor.get.bind(el),
+      set: (v) => {
+        const retval = srcDescriptor.set.call(el, v);
+
+        this.ready(() => this.trigger('sourceset'), true);
+
+        return retval;
+      },
+      configurable: true,
+      enumerable: srcDescriptor.enumerable
+    });
+
+    const oldSetAttribute = el.setAttribute;
+
+    el.setAttribute = (n, v) => {
+      const retval = oldSetAttribute.call(el, n, v);
+
+      if (n === 'src') {
+        this.ready(() => this.trigger('sourceset'), true);
+      }
+
+      return retval;
+    };
+
+    const oldLoad = el.load;
+
+    el.load = () => {
+      const retval = oldLoad.call(el);
+
+      this.ready(() => this.trigger('sourceset'), true);
+
+      return retval;
+    };
+
+    this.on('dispose', () => {
+      if (!el) {
+        return;
+      }
+
+      el.load = oldLoad.bind(el);
+      el.setAttribute = oldSetAttribute.bind(el);
+      Object.defineProperty(el, 'src', srcDescriptor);
+    });
+
+    // observer video tag changes, we do this when tech loads on iOS
+    this.on('sourceset', this.handleSourceSet_);
+  }
+
+  /**
+   * Update the internal source cache based on the `sourceset` event.
+   *
+   * @param {EventTarget~Event} e
+   *        The sourceset event that triggered this function to run.
+   */
+  handleSourceSet_(e) {
```

Comment:
I feel like this function could use some more comments for future maintainers.

Reply:
added a bit more comments